### PR TITLE
HOTT-1315: Prewarm slow to fetch subheadings

### PR DESCRIPTION
**Pull request by willfish**
### Jira link

https://transformuk.atlassian.net/browse/HOTT-1315

### What?

We need to prewarm expensive subheadings as part of our morning ETL job. This will take
the subheading api response times in the worst case down from 12 seconds to 45 milliseconds.

I have added/removed/altered:

- [x] Added ExpensiveHeadingCommodityContextService and specs
- [x] Added PrewarmSubheadingsWorker and specs
- [x] Altered ClearCacheWorker to call PrewarmSubheadingsWorker

### Why?

I am doing this because:

- Massively speed up subheading api for problem cases
- Commodity#children gets called a lot (for every commodity). We can speed it up substantially by introducing the ExpensiveHeadingCommodityContextService (this reduces the time it takes to cache all subheadings from 28 minutes to 2 minutes)
- We can use this new service when constructing headings for the api too and substantially speed up the Heading recaching.

### CODE REVIEW

File: app/models/commodity.rb [hunk 1]
```diff
@@ -166,4 +167,17 @@ def traverse_children(&block)
...
+  def load_children
+    heading.commodities_dataset
+      .eager(:goods_nomenclature_indents, :goods_nomenclature_descriptions)
+      .all
+  end
+
+  def preloaded_children
+    Thread.current[:heading_commodities].try(:fetch, heading_short_code, {})
+  end
+
```
Comment (willfish) [inline app/models/commodity.rb:182]: Remove this extra line

File: spec/workers/prewarm_subheadings_worker_spec.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+      expect(ExpensiveHeadingCommodityContextService).to have_received(:new)
+    end
+  end
+
+  context 'when there are subheadings to populate' do
+    before do
+      GoodsNomenclature.unrestrict_primary_key
+
+      create(:chapter, :with_section, :with_indent, :with_guide, goods_nomenclature_sid: 1, indents: 0, producline_suffix: '80', goods_nomenclature_item_id: '0100000000')       # Live animals
```
Comment (willfish) [inline spec/workers/prewarm_subheadings_worker_spec.rb:29]: Push this tree to a trait?

File: app/models/commodity.rb [hunk 3]
```diff
@@ -166,4 +167,17 @@ def traverse_children(&block)
 
     children.each { |child| child.traverse_children(&block) }
   end
+
+  private
+
+  def load_children
+    heading.commodities_dataset
```
Comment (willfish) [inline app/models/commodity.rb:174]: Move this to the heading model?

File: app/services/expensive_heading_commodity_context_service.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+class ExpensiveHeadingCommodityContextService
```
Comment (willfish) [inline app/services/expensive_heading_commodity_context_service.rb:1]: Add a comment to this

File: spec/workers/clear_cache_worker_spec.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+RSpec.describe ClearCacheWorker, type: :worker do
+  subject(:worker) { described_class.new }
+
+  before do
+    allow(Rails.cache).to receive(:clear)
+    allow(Sidekiq::Client).to receive(:enqueue)
+    allow(Sidekiq::Client).to receive(:enqueue)
```
Comment (alexdesi) [inline spec/workers/clear_cache_worker_spec.rb:7]: A duplicate line here.

**Review by alexdesi — APPROVED** [review]
Nice one!  Just one comment
✅

**Pull request merged**